developer_guides/uuid/index.rst

```diff
@@ -0,0 +1,134 @@
+.. _uuid:
+
+UUID Usage in SOF
+#################
+
+Why UUID is needed
+******************
+
+When developing a new audio signal processing component, we
+traditionally implement the component driver with a new unique component
+type declared together, the painpoints with this method including ABI
+alignment among topology/driver/firmware for each new added components,
+and there could be collision if versions are not aligned.
+
+UUIDs (Universally Unique Identifiers) provide a more scalable and
+collisionn-free way of component identification. UUIDs are currently
+used as the standard interface by all users of the SOF firmware,
+including the tracing subsystem, the topology .m4 files, and the Linux
+topology driver.
+
+Allocating a new UUID for the new added component is mandatory, and
+UUIDs are expected to replace the component type in the near future. For
```

Comment:
no. we need to keep generic types.

Reply:
@kv2019i @plbossart I don't see the mandatory of keeping the component type, e.g. for volume component, it is just like another component which we created a kcontrol for it in the driver side, we use the component id (actually an unique instant id among all kinds of components) for setting/getting the values b/w firmware and driver.

Can you help to share an example which type we have to keep?

Reply:
@keyonjie For example, without component type, how would the kernel driver know that a it should parse "asrc_tokens" when creating the ASRC component, and use "sof_ipc_comp_asrc" struct to create the IPC message? Similarly, how to know when to parse "volume_tokens". For components that have only a blob as a parameter, this is not needed and we can use the generic "effect" component_type. E.g. we could have more robust check e.g. in snd_sof_volume_put() to ensure volume IPC is sent to an actual volume component -- now we rely on the topology to be correct here. If we start exposing more parameters via kcontrols, I'm not sure this scales if kernel is just passing blobs. Is your thinking to move all these to "effects"?

Reply:
@kv2019i we use snd_soc_dapm_type (specified in tplg for each dapm widget) to decide if it is asrc, pga, or effect, we don't use the component type there.

Reply:
@keyonjie Ok, I get your point. But snd_soc_dapm_type is effectively the component type then. I think this should be reflected in rest of the docs. E.g. developer_guides/firmware/component-tutorial/tut-ii-topology.rst does not touch on what is the meaning of the dapm_type, but only covers component_type. After this change, there is a 1:1 mapping between the two and component author needs to pick the dapm-type for the new component.

Reply:
@plbossart Ok with above? I think dapm/comp_type is more of a detail, I'm happy as long as we keep a generic "class" notion in kernel (and not have any per-UUID action in kernel).

Reply:
> @keyonjie Ok, I get your point. But snd_soc_dapm_type is effectively the component type then. I think this should be reflected in rest of the docs. E.g. developer_guides/firmware/component-tutorial/tut-ii-topology.rst does not touch on what is the meaning of the dapm_type, but only covers component_type. After this change, there is a 1:1 mapping between the two and component author needs to pick the dapm-type for the new component.

Yes, that's true, user have to specify a dapm type in .m4 when asking for creating a dapm widget.

Reply:
the PR is updated with this part changed, mark it 'resolved'.